lib/workspace.js

```diff
@@ -190,43 +197,79 @@ module.exports = {
+    // a fair amount of I/O and we don't want to hold a lock during a
+    // potentially long operation. However, we will lock here to ensure
+    // that we don't run into problems if `startup()` was called concurrently
+    // somewhere else:
+    //
+    // - We don't want an interleaving of state transitions like this:
+    //   stopped -> launching -> stopped -> launching
+    // - We don't want multiple hosts trying to assign a host for the same
+    //   workspace at the same time.
+    let shouldAssignHost = false;
+    await sqldb.runInTransactionAsync(async () => {
+      // First, lock the workspace row.
+      const workspaceResults = await sqldb.queryOneRowAsync(sql.select_and_lock_workspace, {
+        workspace_id,
+      });
+      const workspace = workspaceResults.rows[0];
+
+      // If the initial state was `uninitialized`, we should check if it's
+      // still uninitialized. If so, we'll need to perform a state transition.
+      const shouldTransitionToStopped =
+        state === 'uninitialized' && workspace.state === 'uninitialized';
+      if (shouldTransitionToStopped) {
         await module.exports.updateState(workspace_id, 'stopped', 'Initialization complete');
       }
-      await module.exports.updateState(workspace_id, 'launching', 'Assigning workspace host');
-
-      let workspace_host_id = null;
-      let attempt = 0;
-      // eslint-disable-next-line no-constant-condition
-      while (true) {
-        if (attempt > config.workspaceLaunchingRetryAttempts) {
-          throw new Error('Time exceeded to deploy more computational resources');
-        }
-        workspace_host_id = await module.exports.assignHost(workspace_id);
-        if (workspace_host_id != null) {
-          break; // success, we got a host
-        }
-        const t = attempt * config.workspaceLaunchingRetryIntervalSec;
-        await module.exports.updateMessage(
-          workspace_id,
-          `Deploying more computational resources (${t} seconds elapsed)`
-        );
-        await util.promisify(setTimeout)(config.workspaceLaunchingRetryIntervalSec * 1000);
-        attempt++;
+
+      // If the workspace is in the stopped state (or we just transitioned to it),
+      // transition to the launching state.
+
+      if (workspace.state === 'stopped' || shouldTransitionToStopped) {
+        await module.exports.updateState(workspace_id, 'launching', 'Assigning workspace host');
+        shouldAssignHost = true;
       }
-      await module.exports.updateMessage(workspace_id, 'Sending launch command to host');
-      await module.exports.controlContainer(workspace_id, 'init', {
-        useInitialZip,
-      });
+    });
+
+    // Bail out if needed; this should only ever occur if another host is
+    // already trying to assign this host to a workspace.
+    if (!shouldAssignHost) return;
+
+    let workspace_host_id = null;
+    let attempt = 0;
+    // eslint-disable-next-line no-constant-condition
+    while (true) {
```

Comment:
This loop is probably going to cause problems during deploys: if we kill a server, no one else will ever try to assign a workspace to a host. I think there are two ways we could improve this:

- Add a cronjob to periodically try assigning workspaces to hosts. Then if host assignment doesn't immediately succeed after initializing the workspace, allow that cronjob to try performing host assignment instead.
- Switch from a push to a pull-based model, which could take a few forms:
  - Have workspace hosts long-poll PrairieLearn hosts for new workspaces.
  - Do what we do for external graders and push "launch requests" into SQS.